typing/typecore.ml

```diff
@@ -4738,6 +4749,31 @@ and type_cases
       )
       half_typed_cases
   in
+  let cases =
+    let may_backtrack = does_contain_gadt && not !Clflags.principal in
+    if not may_backtrack then mk_cases false else
+    let snap = Btype.snapshot () in
+    let has_equation_escape err =
+      match trace_of_error err with
+        Some tr ->
+          List.exists Ctype.Unification_trace.
+            (function Escape {kind=Equation _} -> true | _ -> false) tr
+      | None -> false
+    in
+    try mk_cases false
+    with Error(_,_,err) when has_equation_escape err ->
+      Btype.backtrack snap;
+      let cases = mk_cases true in
+      let msg =
+        Format.asprintf
+          "@[<v2>@ @[<hov>The return type of this pattern-matching \
+           is ambiguous.@ \
+           Please add a type annotation,@ as the choice of `@[%a@]'@]@]"
+          Printtyp.type_expr ty_res
+      in
+      Location.prerr_warning loc (Warnings.Not_principal msg);
+      cases
```

Comment:
This code is careful to call `mk_cases true` first, and only then show the warning. Can you explain why this is better than showing the warning first and then checking `mk_cases`? Is the idea that we don't want to show the warning if the code is going to fail with an error anyway?

Reply:
> Is the idea that we don't want to show the warning if the code is going to fail with an error anyway?

Yes, you only know that the failure is due to an ambiguity if the change from `false` to `true` fixes the failure. Otherwise the branch could be ill-typed for any other reason (and the warning would then be incorrect)